Add filter to allow third-parties to filter exportable product types

james-allan opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Moves the functionality which generates the array of product types that can be imported/exported to a central function and then adds a filter to allow third-parties to add additional product types. 

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

### How to test the changes in this Pull Request:

1. On its own, this PR shouldn't change any behaviour from the front end. 
2. We need a filter like this in WC Subscriptions so we can register subscription variations as a separate product type. 

See 3970-gh-woocommerce/woocommerce-subscriptions for further details. 

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Dev - Add the 'woocommerce_exporter_product_types' filter to allow third-parties to filter the product types which can be imported and exported.

james-allan committed:
Add filter to allow third-parties to filter exportable product types

Konamiman commented:
Hi @james-allan

> On its own, this PR shouldn't change any behaviour from the front end.

Could you please add some detailed testing steps to verify that that's indeed the case, as well as the changelog entry in the pull request description?

Konamiman commented on includes/admin/class-wc-admin-exporters.php:
```diff
@@ -191,6 +191,30 @@ public function do_ajax_product_export() {
+		/**
+		 * Allow third-parties to filter the exportable product types.
+		 *
+		 * @since 5.0.1
```
This is going into WooCommerce 5.1 so:

```suggestion
		 * @since 5.1.0
```

james-allan commented:
Hi @Konamiman.

Thanks for the feedback. I've fixed that version numbering issue and below I've included steps for how to test the standard exporting flow to confirm this PR doesn't change it. 

> as well as the changelog entry in the pull request description?

I did add a changelog entry to the description however it was inside the quote so I've fixed that.


### Steps to test: 

1. Take a backup of your site's database if you don't want to duplicate products and don't want do delete them manually or via the database after.
2. Go to **Products > Export** 
3. From the product type dropdown all the standard product types should be present. https://d.pr/i/C8Rqtc
    - These options are now passed through the filter added by this PR. 
4. Select a subset of those product types. Just "Simple" for example.
5. Click **Generate CSV**
6. The resulting CSV should contain all your simple products and nothing else.
7. Repeat with a different subset or leave empty to export all products. 

----

1. Go to **Products > Import**
2. Choose the file you've exported. 
3. Click **Continue**
5. On the mapping screen set the ID column to not import. 
6. Continue through at the bottom
7. Once importing is complete. The products would have been added correctly. 

During import, the new filter is used to validate the file's `Type` column. Previously it was just the `wc_get_product_types()` keys + 'variation'.

If you'd like to see how we're using this in Subscriptions, you can find our PR here: 3970-gh-woocommerce/woocommerce-subscriptions

Konamiman reviewed (commented):
I have verified that the import/export works and the filter behaves as expected, so I'm approving.
